test/parallel/test-buffer-tojson.js

```diff
@@ -0,0 +1,35 @@
+'use strict';
+
+require('../common');
+const assert = require('assert');
+const Buffer = require('buffer').Buffer;
+
+{
+  assert.strictEqual(JSON.stringify(Buffer.alloc(0)),
+                     '{"type":"Buffer","data":[]}');
+  assert.strictEqual(JSON.stringify(Buffer.from([1, 2, 3, 4])),
+                     '{"type":"Buffer","data":[1,2,3,4]}');
+}
+
+// issue GH-7849
+{
+  const buf = Buffer.from('test');
+  const json = JSON.stringify(buf);
+  const obj = JSON.parse(json);
+  const copy = Buffer.from(obj);
+
+  assert.deepStrictEqual(buf, copy);
+}
+
+// GH-5110
+{
+  const buffer = Buffer.from('test');
+  const string = JSON.stringify(buffer);
+
+  assert.strictEqual(string, '{"type":"Buffer","data":[116,101,115,116]}');
+
+  assert.deepStrictEqual(buffer, JSON.parse(string, (key, value) => {
+    return value && value.type === 'Buffer' ?
+           Buffer.from(value.data) : value;
+  }));
```

Comment:
Nit: Feel free to ignore if you disagree, but I feel like this might be a little easier to scan and understand if it was written more like this:

```js
  const reviver = function(key, value) {
    return value && value.type === 'Buffer' ? Buffer.from(value.data) : value;
  };
  assert.deepStrictEqual(buffer, JSON.parse(string, reviver));
```